WooCommerce/Classes/ViewRelated/Products/Edit Product/Reviews/ProductReviewsViewController.swift

```diff
@@ -0,0 +1,358 @@
+    ///
+    /// Note: Just because this func runs does not guarantee `didEnter()` or `didLeave()` will run as well.
+    ///
+    func willEnter(state: State) {
+        // no-op for the moment
+    }
+
+    /// Runs whenever the FSM enters a State.
+    ///
+    func didEnter(state: State) {
+        switch state {
+        case .emptyUnfiltered:
+            if isEmpty == true {
+                displayEmptyUnfilteredOverlay()
+            }
+        case .results:
+            break
+        case .placeholder:
+            displayPlaceholderReviews()
+        case .syncing:
+            ensureFooterSpinnerIsStarted()
+        }
+    }
+
+    /// Runs whenever the FSM leaves a State.
+    ///
+    func didLeave(state: State) {
+        switch state {
+        case .emptyUnfiltered:
+            removeAllOverlays()
+        case .results:
+            break
+        case .placeholder:
+            removePlaceholderReviews()
+        case .syncing:
+            ensureFooterSpinnerIsStopped()
+        }
+    }
+
+    /// Should be called before Sync'ing Starts: Transitions to .results / .syncing
+    ///
+    func transitionToSyncingState() {
+        state = isEmpty ? .placeholder : .syncing
+    }
+
+    /// Should be called whenever the results are updated: after Sync'ing (or after applying a filter).
+    /// Transitions to `.results` / `.emptyFiltered` / `.emptyUnfiltered` accordingly.
+    ///
+    func transitionToResultsUpdatedState() {
+        if isEmpty == false {
+            state = .results
+            return
+        }
+
+        state = .emptyUnfiltered
+    }
+}
+
+// MARK: - Sync'ing Helpers
+//
+extension ProductReviewsViewController: SyncingCoordinatorDelegate {
+
+    /// Synchronizes the Orders for the Default Store (if any).
+    ///
+    func sync(pageNumber: Int, pageSize: Int, reason: String? = nil, onCompletion: ((Bool) -> Void)? = nil) {
+        transitionToSyncingState()
+        viewModel.synchronizeReviews(pageNumber: pageNumber, pageSize: pageSize, productID: product.productID) { [weak self] in
+            self?.transitionToResultsUpdatedState()
+            onCompletion?(true)
+        }
+    }
+}
+
+// MARK: - Spinner Helpers
+//
+extension ProductReviewsViewController {
+
+    /// Starts the Footer Spinner animation, whenever `mustStartFooterSpinner` returns *true*.
+    ///
+    private func ensureFooterSpinnerIsStarted() {
+        guard mustStartFooterSpinner() else {
+            return
+        }
+
+        footerSpinnerView.startAnimating()
+    }
+
+    /// Whenever we're sync'ing an Orders Page that's beyond what we're currently displaying, this method will return *true*.
+    ///
+    private func mustStartFooterSpinner() -> Bool {
+        guard let highestPageBeingSynced = syncingCoordinator.highestPageBeingSynced else {
+            return false
+        }
+
+        return viewModel.containsMorePages(highestPageBeingSynced * SyncingCoordinator.Defaults.pageSize)
```

Comment:
non-blocking thought since this in other setup in the codebase as well: this confuses me a bit, I'm not super sure why this is needed since it's called by `ensureFooterSpinnerIsStarted` which already has clear intent to start the footer spinner 😵